src/custom/conversions.rs

```diff
@@ -0,0 +1,463 @@
+                    abs_call,
+                    ctx.iw_context().f64_type(),
+                    "max_flt_pos",
+                )
+            } else {
+                ctx.builder().build_unsigned_int_to_float(
+                    int_max,
+                    ctx.iw_context().f64_type(),
+                    "max_flt",
+                )
+            }?;
+
+            let fabs_call = ctx
+                .builder()
+                .build_call(fabs, &[arg.into()], "flt_pos")?
+                .as_any_value_enum()
+                .into_float_value();
+
+            // TODO: Test this with converting INT_MAX to float to see if we need a cheeky error margin
+            let success = ctx.builder().build_float_compare(
+                FloatPredicate::OLE,
+                fabs_call,
+                flt_int_max,
+                "conversion_valid",
+            )?;
+            let success = ctx.builder().build_int_z_extend(
+                success,
+                ctx.iw_context().i32_type(),
+                "conversion_valid_i32",
+            )?;
+
+            // Perform the conversion unconditionally, which will result
+            // in a poison value if the input was too large. We will
+            // decide whether we return it based on the result of our
+            // earlier check.
+            let trunc_result = if signed {
+                ctx.builder()
+                    .build_float_to_signed_int(arg.into_float_value(), int_ty, "")
+            } else {
+                ctx.builder()
+                    .build_float_to_unsigned_int(arg.into_float_value(), int_ty, "")
+            }?
+            .as_basic_value_enum();
+
+            let placeholder = ctx
+                .iw_context()
+                .struct_type(&[int_ty.as_basic_type_enum()], false)
+                .get_undef();
+            let result_row =
+                ctx.builder()
+                    .build_insert_value(placeholder, trunc_result, 0, "result_row")?;
+            let trunc_err_hugr_val = Value::extension(ConstError::new(
+                2,
+                format!(
+                    "Float value too big to convert to int of given width ({})",
+                    width
+                ),
+            ));
+            let e = emit_value(ctx, &trunc_err_hugr_val)?;
+
+            // Make a struct with both fields (error message and
+            // conversion result) populated, then set the tag to the
+            // to the result of our overflow check.
+            // This should look the same as the appropriate sum instance.
+            let val = sum_ty.get_undef().as_basic_value_enum().into_struct_value();
+            let val = ctx.builder().build_insert_value(val, e, 1, "error val")?;
+            let val = ctx
+                .builder()
+                .build_insert_value(val, result_row, 2, "conversion_result")?;
+            let val = ctx.builder().build_insert_value(val, success, 0, "tag")?;
```

Comment:
I think it's not good to depend on the concrete representation of sums here (might change in the future). Better to rely on `LLVMSumType` instead, e.g. use `sum_ty.build_tag`